### ISSUE TITLE
Data augmentation: training-time transforms for genomic sequences

### ISSUE THREAD

**Issue by smodee** [smodee/pathogen-classifier#2]
## Overview

Implement a configurable data augmentation module for genomic sequences, applied as training-time transforms (not pre-computed to disk). This increases effective dataset size and improves model robustness to the variability seen in real metagenomic surveillance data.

**Architecture:** A PyTorch `Dataset` wrapper class with individually toggleable transforms, applied on-the-fly during training. This avoids storage bloat and allows flexible experimentation.

## Context

After preprocessing (#1), the training set will consist of ~3-5kb sequence chunks labeled by virus family. The raw dataset has 4,791 sequences (fewer after dedup), which after chunking will increase in count but still benefits from augmentation — especially for underrepresented families (Poxviridae, Arenaviridae).

The augmentations are motivated by a real-world deployment scenario: classifying viral fragments from **wastewater metagenomic surveillance** (an ECDC priority). Real samples have variable read lengths, sequencing errors, strand ambiguity, and partial coverage.

### 1. Reverse complement
- **What:** Replace sequence with its reverse complement (A↔T, C↔G)
- **Why:** DNA is double-stranded; the virus family is identical regardless of which strand is sequenced. This is standard practice and a free 2x effective data multiplier
- **Probability:** 0.5 (applied to half of training samples)
- **Implementation:** Simple string operation: reverse the sequence, swap A↔T and C↔G

### 2. Random subsequence cropping
- **What:** Extract a random contiguous subsequence shorter than the full chunk
- **Why:** Simulates variable-length metagenomic fragments. The model learns to classify from partial information
- **Parameters:** `min_crop_ratio` (e.g., 0.5 = crop to at least 50% of chunk length)
- **Probability:** 0.3
- **Implementation:** Random start position, random length ≥ min_crop_ratio × chunk_length

### 3. Simulated sequencing noise
- **What:** Introduce base substitution errors matching real sequencing error profiles
- **Why:** Real sequencing data has characteristic error rates. Training with noise makes the model robust to imperfect input
- **Parameters:**
  - `illumina_mode`: ~0.1% random substitution rate (each base has small chance of being replaced with a random different base)
  - `nanopore_mode`: ~3-5% error rate, biased toward insertions and deletions (indels)
- **Probability:** 0.2
- **Implementation:** Per-base random substitution. For nanopore mode, also randomly insert or delete single bases

### 4. N-masking / random base dropout
- **What:** Randomly replace bases with 'N' (unknown)
- **Why:** Simulates low-quality regions in real sequencing data. Similar to dropout regularization or image occlusion
- **Parameters:** `mask_rate` (e.g., 0.02 = mask 2% of bases)
- **Probability:** 0.2
- **Implementation:** Random positions selected, bases replaced with 'N'

### 5. K-mer shuffling within windows
- **What:** Divide sequence into small windows (e.g., 50bp), randomly shuffle the order of k-mers within each window
- **Why:** Destroys fine-grained positional motifs while preserving local nucleotide composition. Forces the model to learn compositional features rather than memorizing exact motif positions
- **Parameters:** `window_size` (e.g., 50bp), `k` (e.g., 3 for trinucleotides)
- **Probability:** 0.1 (use sparingly — this is destructive)
- **Implementation:** Split window into k-mers, shuffle, rejoin

### Mixup (embedding-space interpolation)
- **What:** Interpolate between two sequences of the same class in the model's embedding space (not in raw sequence space)
- **Why:** Proven regularization technique in image classification, effective for genomic models too
- **Why deferred:** Requires a pre-trained encoder to produce embeddings. Implement after the base model is trained — can be added as a training-time augmentation in a second training round
- **Reference:** Zhang et al., "mixup: Beyond Empirical Risk Minimization" (2018)

## Implementation design

```
src/augmentation.py
```

### Core class structure

```python
class GenomicAugmentationDataset(torch.utils.data.Dataset):
    """
    Wraps a base dataset and applies configurable sequence augmentations.
    
    Usage:
        base_dataset = SequenceDataset('data/processed/train.csv')
        augmented = GenomicAugmentationDataset(
            base_dataset,
            reverse_complement=True,   # p=0.5
            random_crop=True,          # p=0.3, min_ratio=0.5
            sequencing_noise='illumina',  # p=0.2
            n_masking=True,            # p=0.2, rate=0.02
            kmer_shuffle=False,        # off by default
        )
        loader = DataLoader(augmented, batch_size=32, shuffle=True)
    ```

### CLI for previewing augmentations

```bash
# Preview augmented samples (no training, just visualization)
python src/augmentation.py --input data/processed/train.csv --preview 10
```

This should print 10 original vs augmented sequence pairs with the applied transforms labeled, so the user can verify augmentations look reasonable.

## Ordering constraints

- **Chunking happens BEFORE augmentation** (in preprocessing, #1)
- **Augmentation happens AFTER train/val/test split** (training set only — never augment val/test)
- Augmentations are applied **on-the-fly** during training, not written to disk
- Multiple augmentations can be composed (e.g., reverse complement + noise on the same sample)

## Existing code reference

- `src/data_preprocessing.py` (from #1) — produces the chunked train/val/test CSVs that this module wraps
- `src/data_exploration.py` — `compute_sequence_metrics()` can validate augmented data quality
- `requirements.txt` — torch is already included

## Definition of done

- [ ] `src/augmentation.py` with `GenomicAugmentationDataset` class
- [ ] All 5 augmentations implemented and individually toggleable
- [ ] CLI preview mode for visual validation
- [ ] Unit tests in `tests/test_augmentation.py` verifying each transform preserves sequence validity (only valid nucleotides + N, correct length ranges, label preservation)
- [ ] Committed and pushed to `main`

## Dependencies

- **Blocked by:** #1 (preprocessing — needs chunked train CSV)
- **Blocks:** model training (this wraps the training dataset)
